tidy3d/plugins/design/method.py

```diff
@@ -23,335 +20,780 @@ class Method(Tidy3dBaseModel, ABC):
+                boundary_dict[param.name] = (
+                    0,
+                    len(param.allowed_values) - 1,
+                )  # -1 as ints are starting from 0
+            else:
+                boundary_dict[param.name] = param.span
+
+        # Params must be converted back to ints for BO to learn from them
+        invert_param_converter = {
+            param_name: {val: key for key, val in param_values.items()}
+            for param_name, param_values in param_converter.items()
+        }
+
+        # Fn can be defined here to be a combined func of pre, run_batch, post for BO to use
+        utility = UtilityFunction(kind=self.acq_func, kappa=self.kappa, xi=self.xi)
+        opt = BayesianOptimization(
+            f=run_fn, pbounds=boundary_dict, random_state=self.rng_seed, allow_duplicate_points=True
+        )
+
+        # Run variables
+        arg_list = []
+        total_aux_out = []
+        result = []
+        for _ in range(self.initial_iter):
+            next_point = opt.suggest(utility)
+            self._force_int(next_point, parameters)
+            self._handle_param_convert(param_converter, [next_point])
+            arg_list.append(next_point)
+
+        init_output, aux_out = self._extract_output(run_fn(arg_list))
+        self._flatten_and_append(aux_out, total_aux_out)
+
+        for next_point, next_out in zip(arg_list, init_output):
+            result.append(next_out)
+            self._handle_param_convert(invert_param_converter, [next_point])
+            opt.register(params=next_point, target=next_out)
+
+        # Handle subsequent iterations sequentially as BayOpt package does not allow for batched non-random predictions
+        for _ in range(self.n_iter):
+            next_point = opt.suggest(utility)
+            self._force_int(next_point, parameters)
+            self._handle_param_convert(param_converter, [next_point])
+            next_out, aux_out = self._extract_output(run_fn([next_point]))
+            result.append(next_out[0])
+            self._flatten_and_append(aux_out, total_aux_out)
+
+            # ParameterAny values in next_point need to be converted back for the optimizer
+            self._handle_param_convert(invert_param_converter, [next_point])
+            opt.register(params=next_point, target=next_out[0])
+
+        if console is not None:
+            console.log(
+                f"Best Result: {opt.max['target']}\n"
+                f"Best Parameters: {' '.join([f'{param}: {val}' for param, val in opt.max['params'].items()])}\n"
+            )
+
+        # Output fn_args from the BO.opt object - getting results in situ as opt changes type to float
+        fn_args = []
+        for output in opt.res:
+            self._handle_param_convert(param_converter, [output["params"]])
+            fn_args.append(output["params"])
+
+        return fn_args, result, total_aux_out, opt
+
+
+class MethodGenAlg(MethodOptimize, ABC):
```

Comment:
will be interesting to see how these look in the docs API reference. Maybe ask Dario to help you compile that locally and just do a read through before we release it.